### PR TITLE
chore(ci): record benchmarks parameters to be stored in database

### PR DESCRIPTION
**Pull request by soonum**
This is done to comply with the new Zama benchmark standard. Exhaustive parameters list is stored so once it's parsed and send to database, one can easily filter results on such parameters in visualization tool.

### CODE REVIEW

**Comment by soonum** [comment]
Guys I need your review even if we are in draft phase since next step is pretty much modify the code in a somewhat copy/paste fashion to the rest of benched functions.

**Review by IceTDrinker — COMMENTED** [review]
We saw together how to fix the pcc stuff

**Review by IceTDrinker — COMMENTED** [review]
Looks generally ok, I don't know about the python script like this, lots of changes and the input data structure is not clear to me

**Comment by tmontaigu** [comment]
Id also say it looks ok, maybe if the only purpose of the CryptoRecordParameters is to be created then dumped to json
```rust
        let mut params_record =
            utilities::BenchmarkParametersRecord::from_params(param, display_name);
        params_record.write(&bench_id);
```

then I would make the utilities module only expose one function
```rust
pub fn write_as_json<T: Into<CryptoRecordParameters>>(params: T, bench_id: &str, display_name) { ... } 
```

**Comment by soonum** [comment]
I like the idea very much ! I've just one question what about the future benchmarks (integer to be added, as well as CRT decomposition) that requires a different BenchmarkRecordParameters ? How can I change the default values sets in the associated function `from_params()` ?

**Comment by soonum** [comment]
Now I think this is done for tfhe-rs part. I need to implement handling in Slab side and then test it, that why it's not ready to merge.
Otherwise it will simply break the benchmarks.

File: tfhe/benches/utilities.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
+use serde::Serialize;
+use std::fs;
+use std::path::PathBuf;
+#[cfg(feature = "boolean")]
+use tfhe::boolean::parameters::BooleanParameters;
+use tfhe::core_crypto::prelude::*;
+// #[cfg(feature = "shortint")]
+// use tfhe::shortint::keycache::NamedParam;
```
Comment (IceTDrinker) [inline tfhe/benches/utilities.rs:8]: comment

File: tfhe/benches/utilities.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+#[derive(Serialize)]
+struct BenchmarkParametersRecord {
+    display_name: String,
+    crypto_parameters_alias: String,
+    crypto_parameters: CryptoParametersRecord,
+    message_modulus: Option<usize>,
+    carry_modulus: Option<usize>,
+    ciphertext_modulus: usize,
+    polynomial_multiplication: String, // TODO Can we do an Enum ?
```
Comment (IceTDrinker) [inline tfhe/benches/utilities.rs:113]: we can likely do an enum yes

File: .github/workflows/shortint_benchmark.yml [hunk 3]
```diff
@@ -123,13 +123,15 @@ jobs:
...
+          echo "Computing HMac on results file"
+          SIGNATURE="$(slab/scripts/hmac_calculator.sh ${{ env.RESULTS_FILENAME }}.gz '${{ secrets.JOB_SECRET }}')"
           echo "Sending results to Slab..."
           curl -v -k \
           -H "Content-Type: application/json" \
           -H "X-Slab-Repository: ${{ github.repository }}" \
           -H "X-Slab-Command: store_data" \
           -H "X-Hub-Signature-256: sha256=${SIGNATURE}" \
-          -d @${{ env.RESULTS_FILENAME }} \
+          -d @${{ env.RESULTS_FILENAME }}.gz \
```
Comment (IceTDrinker) [inline .github/workflows/shortint_benchmark.yml:136]: could maybe have a variable containing the name with the .gz, would avoid errors

**Review by IceTDrinker — APPROVED** [review]
Looks good to my tired eyes 😪

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 4]
```diff
@@ -92,7 +62,51 @@ fn benchmark_parameters<Scalar: Numeric>() -> Vec<(String, BenchmarkPbsParameter
     }
 }
 
+fn multi_bit_benchmark_parameters<Scalar: Numeric>(
+) -> Vec<(String, (CryptoParametersRecord, LweBskGroupingFactor))> {
+    if Scalar::BITS == 64 {
+        vec![
+            (
+                "4_bits_multi_bit_group_2".to_string(),
```
Comment (IceTDrinker) [inline tfhe/benches/core_crypto/pbs_bench.rs:70]: still unsure about this, do you want to add the multi bit in the same PR ?

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 5]
```diff
@@ -92,7 +62,51 @@ fn benchmark_parameters<Scalar: Numeric>() -> Vec<(String, BenchmarkPbsParameter
...
+                (
+                    CryptoParametersRecord {
+                        lwe_dimension: Some(LweDimension(724)),
+                        lwe_modular_std_dev: Some(StandardDev(1.58705e-10)),
+                        pbs_base_log: Some(DecompositionBaseLog(1)),
+                        pbs_level: Some(DecompositionLevelCount(18)),
```
Comment (IceTDrinker) [inline tfhe/benches/core_crypto/pbs_bench.rs:76]: params are wrong, it's my bad

File: tfhe/benches/core_crypto/pbs_bench.rs (hunk 4 above)
Comment (soonum) [reply tfhe/benches/core_crypto/pbs_bench.rs:70]: Yes we'll keep your initial commit so we can get two birds with one stone.

File: tfhe/benches/integer/bench.rs [hunk 6]
```diff
@@ -19,15 +23,16 @@ use tfhe::shortint::parameters::{
...
 impl Default for ParamsAndNumBlocksIter {
     fn default() -> Self {
-        const PARAMS: [tfhe::shortint::Parameters; 3] = [
+        // FIXME One set of parameter is tested since we want to benchmark only quickest operations.
+        const PARAMS: [tfhe::shortint::Parameters; 1] = [
             PARAM_MESSAGE_2_CARRY_2,
-            PARAM_MESSAGE_3_CARRY_3,
-            PARAM_MESSAGE_4_CARRY_4,
+            // PARAM_MESSAGE_3_CARRY_3,
+            // PARAM_MESSAGE_4_CARRY_4,
```
Comment (IceTDrinker) [inline tfhe/benches/integer/bench.rs:35]: remove comments ?

File: tfhe/benches/integer/bench.rs [hunk 7]
```diff
@@ -396,12 +482,33 @@ criterion_group!(
...
-    unchecked_arithmetic_operation,
-    unchecked_scalar_arithmetic_operation,
-    misc,
+    fast_integer_benchmarks,
+    // smart_arithmetic_operation,
+    // smart_arithmetic_parallelized_operation,
+    // smart_scalar_arithmetic_operation,
+    // smart_scalar_arithmetic_parallel_operation,
+    // unchecked_arithmetic_operation,
+    // unchecked_scalar_arithmetic_operation,
```
Comment (IceTDrinker) [inline tfhe/benches/integer/bench.rs:512]: comments to remove ?

**Pull request merged**